### PR TITLE
New API endpoint to set project as busy / not busy (1)

### PR DESCRIPTION
**Pull request by inaod568**
# Description

In order to avoid multiple concurrent actions on the same project, this endpoint allows us to set a project to busy --> block other actions for the time being.

- [ ] Summary of the changes and the related issue
- [x] Relevant motivation and context
- [ ] Any dependencies that are required for this change

Fixes DDS-1334

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

Please delete options that are not relevant.

- [x] Any dependent changes have been merged and published in downstream modules
- [x] Rebase/merge the branch which this PR is made to
- [ ] Changes to the database schema: A new migration is included in the PR
- [ ] Product Owner / Scrum Master: This PR is made to the `master` branch and I have updated the [version](../dds_web/version.py)
- [ ] I am bumping the major version (e.g. 1.x.x to 2.x.x) and I have made the corresponding changes to the CLI version

## Formatting and documentation

- [ ] I have added a row in the [changelog](../CHANGELOG.md)
- [x] The code follows the style guidelines of this project: Black / Prettier formatting
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Tests

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

<a href="https://gitpod.io/#https://github.com/ScilifelabDataCentre/dds_web/pull/1266"></a>

### CODE REVIEW

File: dds_web/api/project.py [hunk 1]
```diff
@@ -867,3 +869,41 @@ def give_project_access(project_list, current_user, user):
...
+
+        # Get busy or not busy
+        set_to_busy = flask.request.json.get("busy")
+        if set_to_busy is None:
+            raise DDSArgumentError(message="Are you trying to set the project as busy or not busy?")
+
+        if set_to_busy:
+            # Check if project is busy
+            if project.busy:
+                raise ProjectBusyError(message="The project is already busy, cannot proceed.")
```
Comment (inaod568) [inline dds_web/api/project.py:894]:
```suggestion
                return {"ok": False, "message": "The project is already busy, cannot proceed."}
```

File: dds_web/api/project.py [hunk 2]
```diff
@@ -867,3 +869,41 @@ def give_project_access(project_list, current_user, user):
...
+
+            # Set project as busy
+            project.busy = True
+        else:
+            # Check if project is not busy
+            if not project.busy:
+                raise ProjectBusyError(message="The project is not busy, cannot proceed.")
```
Comment (inaod568) [inline dds_web/api/project.py:901]:
```suggestion
                return {"ok": False, "message": "The project is already not busy, cannot proceed."}
```

File: tests/test_project_busy.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    # Unit Personnel
+    token = tests.UserAuth(tests.USER_CREDENTIALS["unituser"]).token(client)
+    response = client.put(
+        tests.DDSEndpoint.PROJECT_BUSY,
+        headers=token,
+    )
+    assert response.status_code == http.HTTPStatus.BAD_REQUEST
+    assert "Required data missing" in response.json.get("message")
+
+    # Unit Personnel
```
Comment (inaod568) [inline tests/test_project_busy.py:51]:
```suggestion
    # Unit Personnel
```
```suggestion
    # Unit Admin
```

File: tests/test_project_busy.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    """busy bool required."""
+    token = tests.UserAuth(tests.USER_CREDENTIALS["projectowner"]).token(client)
+    response = client.put(
+        tests.DDSEndpoint.PROJECT_BUSY,
+        headers=token,
+        query_string=proj_query,
+        json={"something": "notabool"},
+    )
+    assert response.status_code == http.HTTPStatus.BAD_REQUEST
+    assert "Are you trying to set the project as busy or not busy?" in response.json.get("message")
```
Comment (inaod568) [inline tests/test_project_busy.py:80]:
```suggestion
    assert "Missing information about setting busy or not busy." in response.json.get("message")
```

File: dds_web/api/project.py [hunk 5]
```diff
@@ -867,3 +869,42 @@ def give_project_access(project_list, current_user, user):
...
+        """Set project to busy / not busy."""
+        # Get project ID, project and verify access
+        project_id = dds_web.utils.get_required_item(obj=flask.request.args, req="project")
+        project = dds_web.utils.collect_project(project_id=project_id)
+        dds_web.utils.verify_project_access(project=project)
+
+        # Get busy or not busy
+        set_to_busy = flask.request.json.get("busy")
+        if set_to_busy is None:
+            raise DDSArgumentError(message="Are you trying to set the project as busy or not busy?")
```
Comment (inaod568) [inline dds_web/api/project.py:889]:
```suggestion
            raise DDSArgumentError(message="Missing information about setting busy or not busy.")
```

File: tests/test_project_busy.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    """Set a busy project as busy."""
+
+    token = tests.UserAuth(tests.USER_CREDENTIALS["unitadmin"]).token(client)
+    response = client.put(
+        tests.DDSEndpoint.PROJECT_BUSY,
+        headers=token,
+        query_string=busy_proj_query,
+        json={"busy": True},
+    )
+    # assert response.status_code == http.HTTPStatus.OK
```
Comment (inaod568) [inline tests/test_project_busy.py:122]:
```suggestion
    assert response.status_code == http.HTTPStatus.OK
```

File: tests/test_project_busy.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    assert "The project is already busy, cannot proceed." in response.json.get("message")
```
Comment (inaod568) [inline tests/test_project_busy.py:123]: After this we can also add a new test with the same test except regarding "not busy". So basically just copy the test above and change to not busy.

**Comment by inaod568** [comment]
@valyo Also add an item to the changelog, forgot to mention that in the review

**Review by valyo — APPROVED** [review]
all good now

**Pull request merged**